M2/Macaulay2/packages/Hadamard.m2

```diff
@@ -0,0 +1,441 @@
+
+hadamardProduct=method()
+hadamardProduct(Ideal,Ideal):=(I,J)->(hadProdOfVariety(I,J))
+hadamardProduct(List,List):=(X,Y)->(hadProdListsOfPoints(X,Y))
+hadamardProduct(List):=(L)->(hadamardMult(L));
+
+
+------------------Hadamard product ends ------------------
+
+
+
+
+-----------------------new results-------------
+idealOfProjectivePoints=method()
+idealOfProjectivePoints(List,Ring):=(L,R)->(
+    if not uniform L then 
+     (return "error: entries should be in the same class");
+    MP:=transpose pointsToMatrix L; 
+    return ideal projectivePointsByIntersection(MP,R)
+    )
+
+--------Terracini lemma------
+
+
+
+
+
+
+-----------------------------------------------
+beginDocumentation()
+
+doc ///
+     Key
+       Hadamard
+     Headline
+       a package to study Hadamard products of varieties.
+///
+
+doc ///
+    Key
+    	Point
+    Headline
+    	a new type for points in projective space
+    Description
+    	Text
+	   A point in projective space is represented as an object in the class @TO Point@. An element of this class is a @TO BasicList@.
+///
+
+doc ///
+    Key
+       (symbol *, Point, Point)
+    Headline
+    	entrywise product of two projective points
+    Usage
+    	p * q
+    Inputs
+    	p:Point
+	q:Point
+    Outputs
+    	:Point
+    Description
+    	Example
+	    p = point {1,2,3};
+	    q = point {-1,2,5};
+	    p * q
+	Text
+	    Note that this operation is not always well-defined in projective space, 
+	    e.g., the Hadamard product of the points $[1:0]$ and $[0:1]$ is not well-defined
+///
+
+doc ///
+    Key
+    	(symbol ==, Point, Point)
+    Headline
+    	check equality of two projective points
+    Usage
+    	p == q
+    Inputs
+    	p:Point
+	q:Point
+    Outputs
+    	:Boolean
+    Description
+    	Example
+	    p = point {1,1};
+	    q = point {2,2};
+	    p == q
+///
+
+doc ///
+    Key
+    	point
+        (point, VisibleList)
+    Headline
+        consructs a projective point from the list (or array) of coordinates.
+    Usage
+        point(L)
+    Inputs
+        L:List 
+	   or @TO2{Array, "array"}@
+	   or @TO2{VisibleList, "visible list"}@
+    Outputs
+        :Point
+    Description
+        Example
+            point {1,2,3}
+            point [1,4,6]
+///
+
+doc ///
+    Key
+    	hadamardProduct
+    Headline
+        computes the Hadamard product of varieties
+///
+
+doc ///
+    Key
+	(hadamardProduct, Ideal, Ideal)
+    Headline
+         Hadamard product of two homogeneous ideals
+    Usage
+        hadamardProduct(I,J)
+    Inputs
+        I:Ideal
+	    (homogeneous)
+	J:Ideal
+	    (homogeneous)
+    Outputs
+         :Ideal
+    Description
+        Text
+            Given two projective varieties $X$ and $Y$, their Hadamard product is defined as the
+	    Zariski closure of the set of (well-defined) entrywise products of pairs of points in the cartesian 
+	    product $X \times Y$. This can also be regarded as the image of the Segre product of $X \times Y$
```

Comment:
Not only should the varieties be projective, but they should be given as subvarieties of projective space explicitly, so the coordinates are determined.

Reply:
OK! fixed.